### DISCUSSION TITLE
Ability to store small amount of data in serverless functions provided by vercel

### DISCUSSION THREAD

**Discussion by ghost** [vercel/next.js#36806] [Ideas]
### Describe the feature you'd like to request

# My Use Case
I was writing an app which uses Next Js Api routes. These Api routes are deployed by Vercel in form of serverless function. The Functions boot up on request and shut downs after a specific amout of time. My app needs a small amount of storage (So Small to consider a database, Just a few key-value pairs carrying same value for everyone). I need the storage like so that I can read and write that storage anytime and it have to hold up that data till the next deployment. 
# The Problem
I initially tried using a JSON file and fs module but file read write is not allowed in serverless functions.
Then i tried using `global` variable. It works good in local environment but in serverless function it stays for sometime and gets reset after sometime.
My project needs only 5-6 key value pair storage and bringing an database would be an overkill.
# My Request
A Simple Key Value Paired database like `localStorage` to hold data during the lifespan of an deployment. (Wouldn't mind if the data gets reset after each deployment)

### Describe the solution you'd like

I think a new module like `next/database` should be added which can hold some data after each deployment.
Keys for storing data may be decided initially with their initial Value. Dynamic Keys may be also considered.
Also we need two function to read and write data.
I think Vercel's infrastructure also need to be modified to hold this data.

### Describe alternatives you've considered

I am not able to figure out any permanent solution till now.
I have another github repo setup where i have hosted an file and change that according to my need and fetch that data in api routes.
It works but is very painfull.
Data can only be read through api routes and writes are to be done manually.

**Comment by balazsorban44** [comment]
Why not just use a simple object? Or environment variables?

If the data is user-specific, have you considered cookies?

**Reply by ghost** [reply]
I want to store the data server side and not expose to end user.
The problem with object is that the object will be destroyed as soon as the serverless function sleeps and there is no way to get that back in next boot.
Environmental Variables will work till the function is running, on next boot it will pick env variables from my configured settings. So it can't be changed through next api routes.

**Comment by GabenGar** [comment]
"Just a small amount of keys" doesn't mean anything, since the value for a key can be several megabytes.
"Database" doesn't always mean something humongous. SQLite runs on embedded devices, so it's more than fitting for your usecase. Or Redis.
Or even better rethink if you really need to store the state between requests. If you do, chances are you need a database anyway, otherwise that potential storage will have to implement its own concurrency model, which definitely won't reduce complexity.

**Reply by ghost** [reply]
> "Just a small amount of keys" doesn't mean anything, since the value for a key can be several megabytes. "Database" doesn't always mean something humongous. SQLite runs on embedded devices, so it's more than fitting for your usecase. Or Redis. Or even better rethink if you really need to store the state between requests. If you do, chances are you need a database anyway, otherwise that potential storage will have to implement its own concurrency model, which definitely won't reduce complexity.

If You Are Right Can you explain the use of sqlite in next.js api with out creating a separate backend

**Reply by GabenGar** [reply]
API callbacks and `getServerSideProps()`/`getStaticProps()` run in the server context so you just call the db functions within them.

**Comment by ogbotemi-2000** [comment]
Hello @ghost, I had similar issues when I migrated the backend of my Node.js web app from Railway to Vercel.
Here is the guide I wrote to explain the fix I developed around this limitation:
https://dev.to/ogbotemi2000/persist-data-in-vercelnextjs-serverless-backends-1h70

TL:DR;
https://github.com/vercel/next.js/discussions/15054#discussioncomment-9821324
